runners/direct-java/src/main/java/org/apache/beam/runners/direct/MultiStepCombine.java

```diff
@@ -472,6 +474,7 @@ public MergeAccumulatorsAndExtractOutputEvaluator(
           ctxt.createBundle(
               (PCollection<KV<K, OutputT>>)
                   Iterables.getOnlyElement(application.getOutputs().values()));
+      combineFn.setup();
```

Comment:
I would prefer to not do real work in a constructor. Can this be done with lazy init in processElement?

Have you taken a look at how DoFn setup and teardown are managed with a loading cache in the directrunner?